Cython/Compiler/ExprNodes.py

```diff
@@ -7374,7 +7384,13 @@ def generate_result_code(self, code):
                         '%s'
                     '}' % (self.result(), code.error_goto(self.pos)))
         elif self.entry.is_cpp_optional and self.initialized_check:
-            unbound_check_code = self.type.cpp_optional_check_for_null_code(self.result())
+
+            if self.entry and self.entry.is_cpp_optional and not self.is_target:
+                assert not self.is_temp  # calculate_access_code only makes sense for non-temps
+                undereferenced_result = self.calculate_access_code()
+            else:
+                undereferenced_result = self.result()
```

Comment:
`self.entry.is_cpp_optional` is already checked just above.
```suggestion
            if self.is_target:
                undereferenced_result = self.result()
            else:
                assert not self.is_temp  # calculate_access_code() only makes sense for non-temps
                undereferenced_result = self.calculate_access_code()
```